lib/Basic/FileSystem.cpp

```diff
@@ -29,6 +35,82 @@ namespace {
   };
 } // end anonymous namespace
 
+std::error_code swift::atomicallyWritingToFile(
+    StringRef outputPath, bool binaryMode,
+    llvm::function_ref<void(llvm::raw_pwrite_stream &)> action) {
```

Comment:
I wanted to check the correctness of this function for my review, but it is so long that I doubt my ability to do so.
If you were to break it up into smaller pieces, I could review it with confidence.